### PR TITLE
feat: google ads identify call support for enhanced conversion

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## PR Description

Google ads identify call support for enhanced conversion

## Linear task (optional)

Resolves INT-2073
Linear task https://linear.app/rudderstack/issue/INT-2073/google-ads-device-mode-add-user-properties

## Cross Browser Tests

Please confirm you have tested for the following browsers:

- [ ] Chrome
- [ ] Firefox
- [ ] IE11

## Sanity Suite

- [ ] All sanity suite test cases pass locally

## Security

- [ ] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

## Summary by CodeRabbit

- **New Features**
  - Enhanced Google Ads integration with user identification functionality for sending user data like email, phone, and address.
  - Added `allowIdentify` configuration option for enabling user identification.

- **Tests**
  - Implemented additional tests for Google Ads identification feature, encompassing scenarios with and without user traits.

### CODE REVIEW

File: packages/analytics-js-integrations/src/integrations/GoogleAds/browser.js [hunk 1]
```diff
@@ -48,6 +50,7 @@ class GoogleAds {
     this.dynamicRemarketing = config.dynamicRemarketing;
     this.allowEnhancedConversions = config.allowEnhancedConversions || false;
     this.v2 = config.v2 || true;
+    this.allowIdentify = config.allowIdentify ?? true;
```
Comment (Gauravudia) [inline packages/analytics-js-integrations/src/integrations/GoogleAds/browser.js:53]: shouldn't it be false by default?

File: packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js [hunk 2]
```diff
@@ -127,9 +127,51 @@ function newCustomerAcquisitionReporting(properties) {
...
+    }
+    if (isDefinedAndNotNull(traits.city)) {
+      payload.address.city = traits.city;
+    }
+    if (isDefinedAndNotNull(traits.street)) {
+      payload.address.street = traits.street;
+    }
+    if (isDefinedAndNotNull(traits.region)) {
+      payload.address.region = traits.region;
+    }
```
Comment (Gauravudia) [inline packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js:160]:
```suggestion
    if (isDefinedAndNotNull(traits.state)) {
      payload.address.region = traits.state;
    }
```

Comment (Gauravudia) [reply packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js:160]: refer :- https://www.rudderstack.com/docs/event-spec/standard-events/identify/#identify-traits

**Comment by anantjain45823** [comment]
@yashasvibajpai we might need to consider trimming before hashing of user properties 
Cc @utsabc

**Comment by Gauravudia** [comment]
@yashasvibajpai What happens if the data is already hashed? Does it make sense to expose a toggle in the UI for this?
Let's say for email:

- If the data is not hashed, we will send the email under the `email` property to Google Ads. Then Google Ads will hash the data on their end.
- If the data is already hashed, we will send the email under the `sha256_email_address` property.

**Comment by Gauravudia** [comment]
@yashasvibajpai I don't see this condition has been satisfied. Shouldn't we throw the error acc. to this?
```
Note: At least one of the following fields must be provided:
- Email (preferred)
- Address (first name, last name, postal code, and country are required). You can optionally provide street address, city, and region as additional match keys.
- A phone number can also be provided along with an email or full name and address
```

https://support.google.com/google-ads/answer/13258081#zippy=%2Cvalidate-your-implementation-using-chrome-developer-tools%2Cfind-enhanced-conversions-fields-on-your-conversion-page%2Cidentify-and-define-your-enhanced-conversions-fields

**Comment by yashasvibajpai** [comment]
> @yashasvibajpai What happens if the data is already hashed? Does it make sense to expose a toggle in the UI for this? Let's say for email:
> 
> * If the data is not hashed, we will send the email under the `email` property to Google Ads. Then Google Ads will hash the data on their end.
> * If the data is already hashed, we will send the email under the `sha256_email_address` property.
> 
> https://support.google.com/google-ads/answer/13258081#zippy=%2Cvalidate-your-implementation-using-chrome-developer-tools%2Cfind-enhanced-conversions-fields-on-your-conversion-page%2Cidentify-and-define-your-enhanced-conversions-fields

The customer is looking for unhashed right now urgently, we can pick this up later @Gauravudia

File: packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js (hunk 2 above)
Comment (yashasvibajpai) [reply packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js:160]: Addressed

File: packages/analytics-js-integrations/src/integrations/GoogleAds/browser.js (hunk 1 above)
Comment (yashasvibajpai) [reply packages/analytics-js-integrations/src/integrations/GoogleAds/browser.js:53]: Addreseed

**Review by Gauravudia — DISMISSED** [review]

File: packages/analytics-js-integrations/src/integrations/GoogleAds/browser.js [hunk 3]
```diff
@@ -87,6 +90,34 @@ class GoogleAds {
...
+  identify(rudderElement) {
+    if (this.allowIdentify === false) {
+      logger.info(
+        'Please enable identify call toggle in your destination settings to send user data to Google Ads.',
+      );
+      return;
+    }
+    const { message } = rudderElement;
+    const { context } = message;
+    const { traits } = context;
```
Comment (ItsSudip) [inline packages/analytics-js-integrations/src/integrations/GoogleAds/browser.js:102]:
```suggestion
    const { traits } = rudderElement.message?.context;
```

File: packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js [hunk 4]
```diff
@@ -127,9 +127,51 @@ function newCustomerAcquisitionReporting(properties) {
...
+    if (isDefinedAndNotNull(traits.state)) {
+      payload.address.region = traits.state;
+    }
+    if (isDefinedAndNotNull(traits.postalCode)) {
+      payload.address.postal_code = traits.postalCode;
+    }
+    if (isDefinedAndNotNull(traits.country)) {
+      payload.address.country = traits.country;
+    }
```
Comment (ItsSudip) [inline packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js:166]: Can you add a check to remove address field from payload when none of the address fields are provided.

**Review by shrouti1507 — DISMISSED** [review]

**Review by ItsSudip — DISMISSED** [review]

**Review by ItsSudip — APPROVED** [review]

File: packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js (hunk 4 above)
Comment (yashasvibajpai) [reply packages/analytics-js-integrations/src/integrations/GoogleAds/utils.js:166]: added

**Pull request merged**